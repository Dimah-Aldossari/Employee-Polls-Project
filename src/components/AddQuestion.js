// AddQuestion.js

import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { _saveQuestion } from '../_DATA';
import { createQuestion } from '../actions/questionsSlice';
import { useNavigate } from 'react-router-dom';

const AddQuestion = () => {
    const dispatch = useDispatch();
    const authUser = useSelector((state) => state.authSlice.user);
    const navigate = useNavigate();

    const [options, setOptions] = useState({
        optionOneText: '',
        optionTwoText: '',
        author: authUser ? authUser.id : null,

    });

    const handleInputChange = (e, option) => {
        setOptions((prevOptions) => ({
            ...prevOptions,
            [option]: e.target.value,
        }));
    };

    const handleCreateQuestion = async () => {
        try {
            const newQuestion = await _saveQuestion({
                optionOneText: options.optionOneText,
                optionTwoText: options.optionTwoText,
                author: authUser, // تغيير هنا
            });
            dispatch(createQuestion(newQuestion));

            // يمكنك إعادة توجيه المستخدم إلى صفحة أخرى بعد إنشاء السؤال
            navigate('/Home');
        } catch (error) {
            console.error('Error creating question:', error);
        }
    };

    return (
        <div className='Create-Question'>
            <h2>Create a New Question</h2>
            <div className='Option'>
                <label
                    data-testid="firstOptionLabel">
                    Option One:
                    <input
                        data-testid="firstOption"
                        className='input'
                        type="text"
                        value={options.optionOneText}
                        onChange={(e) => handleInputChange(e, 'optionOneText')}
                    />
                </label>
            </div>
            <div className='Option'>
                <label
                    data-testid="secondOptionLabel"
                >
                    Option Two:
                    <input
                        data-testid="secondOption"
                        className='input'
                        type="text"
                        value={options.optionTwoText}
                        onChange={(e) => handleInputChange(e, 'optionTwoText')}
                    />
                </label>
            </div>
            <button data-testid="submit-poll" className='btn' onClick={handleCreateQuestion}>Create Question</button>
        </div>
    );
};

export default AddQuestion;
