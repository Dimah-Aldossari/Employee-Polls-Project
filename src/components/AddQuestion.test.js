import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "../Store";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import AddQuestion from "./AddQuestion"

describe("AddQuestion", () => {
    it("should render the component", () => {
        const view = render(
            <Provider store={store}>
                <BrowserRouter>
                    <AddQuestion />
                </BrowserRouter>
            </Provider>
        );
        expect(view).toBeDefined();
        expect(view).toMatchSnapshot();
    });

    it("should display all elements", () => {
        const view = render(
            <Provider store={store}>
                <BrowserRouter>
                    <AddQuestion />
                </BrowserRouter>
            </Provider>
        );

        const firstOptionLabelElement = screen.getByTestId("firstOptionLabel");
        const firstOptionInputElement = screen.getByTestId("firstOption");
        const secondOptionLabelElement = screen.getByTestId("secondOptionLabel");
        const secondOptionInputElement = screen.getByTestId("secondOption");
        const submitButtonElement = screen.getByTestId("submit-poll");

        expect(firstOptionLabelElement.textContent).toBe("Option One:");
        expect(secondOptionLabelElement.textContent).toBe("Option Two:");
        expect(submitButtonElement.textContent).toBe("Create Question");

        fireEvent.change(firstOptionInputElement, { target: { value: 'Texas' } });
        fireEvent.change(secondOptionInputElement, { target: { value: 'New Hampshire' } });
        expect(firstOptionInputElement.value).toBe("Texas");
        expect(secondOptionInputElement.value).toBe("New Hampshire");
    });
});
