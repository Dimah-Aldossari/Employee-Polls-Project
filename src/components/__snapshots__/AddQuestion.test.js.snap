// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddQuestion should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Create-Question"
      >
        <h2>
          Create a New Question
        </h2>
        <div
          class="Option"
        >
          <label
            data-testid="firstOptionLabel"
          >
            Option One:
            <input
              class="input"
              data-testid="firstOption"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="Option"
        >
          <label
            data-testid="secondOptionLabel"
          >
            Option Two:
            <input
              class="input"
              data-testid="secondOption"
              type="text"
              value=""
            />
          </label>
        </div>
        <button
          class="btn"
          data-testid="submit-poll"
        >
          Create Question
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Create-Question"
    >
      <h2>
        Create a New Question
      </h2>
      <div
        class="Option"
      >
        <label
          data-testid="firstOptionLabel"
        >
          Option One:
          <input
            class="input"
            data-testid="firstOption"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="Option"
      >
        <label
          data-testid="secondOptionLabel"
        >
          Option Two:
          <input
            class="input"
            data-testid="secondOption"
            type="text"
            value=""
          />
        </label>
      </div>
      <button
        class="btn"
        data-testid="submit-poll"
      >
        Create Question
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
